src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,109 @@ namespace cryptonote
-        "/sys/class/power_supply/ADP0/online"
-      };
+      // Use the power_supply class http://lxr.linux.no/#linux+v4.10.1/Documentation/power/power_supply_class.txt
+      std::string power_supply_class_path = "/sys/class/power_supply";
 
-      for(const std::string& path : POWER_SUPPLY_STATUS_PATHS)
+      if (epee::file_io_utils::is_directory(power_supply_class_path))
       {
-        if( epee::file_io_utils::is_file_exist(path) )
-        {
-          power_supply_path = path;
-          break;
-        }
-      }
+        std::vector<boost::filesystem::path> paths;
+        epee::file_io_utils::get_directory_paths(power_supply_class_path, paths);
 
-      if( power_supply_path.empty() )
-      {
-        LOG_ERROR("Couldn't find battery/power status file, can't determine if plugged in!");
-        return boost::logic::tribool(boost::logic::indeterminate);;
-      }
+        // Iterate through power_supplies alphabetically.
+        // This ensures that we iterate through potential
+        // AC power supplies first.
+        std::sort(paths.begin(), paths.end());
 
-      std::ifstream power_stream(power_supply_path);
-      if( power_stream.fail() )
-      {
-        LOG_ERROR("failed to open '" << power_supply_path << "'");
-        return boost::logic::tribool(boost::logic::indeterminate);;
+        // Take the first present power supply as truth
+        for (std::vector<boost::filesystem::path>::const_iterator iter(paths.begin()), iter_end(paths.end()); iter != iter_end; ++iter)
+        {
+          std::string power_supply_path = (*iter).string();
+          if (epee::file_io_utils::is_directory(power_supply_path))
+          {
+            std:string power_supply_present = epee::file_io_utils::file_to_string(power_supply_path + "/present");
+            if (!boost::starts_with(power_supply_present, "1")) {
+              LOG_PRINT_L4("Power supply not present at " << power_supply_path);
+              continue;
+            }
+
+            std::string power_supply_type_path = power_supply_path + "/type";
+            if (epee::file_io_utils::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type = epee::file_io_utils::file_to_string(power_supply_type_path);
+
+              if (boost::starts_with(power_supply_type, "Battery"))
+              {
+                std::string power_supply_status_path = power_supply_path + "/status";
+                if (epee::file_io_utils::is_regular_file(power_supply_status_path))
+                {
+                  std::string power_supply_status = epee::file_io_utils::file_to_string(power_supply_status_path);
+
+                  if (boost::starts_with(power_supply_status, "Charging"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else if (boost::starts_with(power_supply_status, "Full"))
+                  {
+                    return boost::logic::tribool(false);
```

Comment:
So you are saying you are on a laptop, plugged into the wall and have no AC power source? That would be very odd and perhaps a system specific issue. My ArchLinux doesn't even have the acpi status files so it's not a science here. The current master code at least checks a few common names for AC sources and I can only assume thats why it is working for the majority. 

I like the fact you wanted to improve it but think you got tangled in a logic problem. Looping all sources and then applying logic is the simple change rather than throwing the toys out the pram and deleting the PR because of peer review which is only trying to help get it as best as possible for the sake of the project.